JS: Make sanitizers no longer block taint inside an object

asgerf opened the pull request:
Taint can now be tracked across sanitizers that don't operate on the tainted value directly, but only on an object containing it. For example:
```js
if (isSafe(x)) {
  sink(x.foo); // NOT OK - 'x' is sanitized, but not 'x.foo'
}
```

This was done through the following change to taint-tracking configurations:
- Taint sources are now associated with the `taint` label by default (previously `data`).
- Sanitizers only block the `taint` label.

The old behavior of completely blocking flow through a label can be achieved by overriding `isBarrier` instead, or using a labeled barrier to block the `data` label.

The flow label mapping `data->taint` is still generated by all taint steps. This ensures backwards compatibility for queries that use `data` as the source (although the change to sanitizers is still in effect).

Note that we have some taint-tracking configurations that relies on `data` being the source, such as HardcodedDataInterpretedAsCode, where we are only interested in data that has gone through some transformation before reaching the sink. So there are use-cases for having `data` being the source, it's just no longer the default.

[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/js/property-barriers_1582386662318) was pretty quiet.

erik-krogh commented on javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll:
```diff
@@ -87,22 +87,26 @@ module TaintTracking {
-    final override predicate isBarrierEdge(DataFlow::Node source, DataFlow::Node sink) {
-      super.isBarrierEdge(source, sink) or
-      isSanitizerEdge(source, sink)
+    override predicate isBarrier(DataFlow::Node node) {
```
You removed the `final override predicate isBarrierEdge(DataFlow::Node source, DataFlow::Node sink)`. 

I think we should keep the override for consistency with the labeled variant of `isBarrierEdge` (just without calling `isSanitizerEdge`). 
E.g. 
```
final override predicate isBarrierEdge(DataFlow::Node source, DataFlow::Node sink) {	
  super.isBarrierEdge(source, sink)
}
```

asgerf replied:
Hm, so the idea was that you could override the non-labeled version of `isBarrier` and `isBarrierEdge` to block all labels. To do the same with the labeled versions requires making a cartesian product of all flow labels, which can be expensive when there are also many barrier nodes/edges.

For example, suppose you block all nodes that occur in client-side code, and there are 8 flow labels, the factor 64 is enough to matter.

erik-krogh commented on javascript/ql/test/library-tests/TaintTracking/object-bypass-sanitizer.js:
```diff
@@ -1,10 +1,20 @@
+    sink(sanitizer_id(x)); // OK
+    sink(sanitizer_id(x.foo)); // OK
+    sink(sanitizer_id(x).foo); // NOT OK
 }
```
Could you duplicate `f` into two test methods? 
Such that we have one function that is invoked like: `f(source());`, and another function that is invoked like `f({foo: source()});`. 
That would make it easier to understand what is supposed to happen for each case.

erik-krogh commented on change-notes/1.24/analysis-javascript.md:
```diff
@@ -9,7 +9,10 @@
 * Imports with the `.js` extension can now be resolved to a TypeScript file,
   when the import refers to a file generated by TypeScript.
 
-* Imports that rely on path-mappings from a `tsconfig.json` file can now be resolved.
```
Did you remove the `* Imports that rely on path-mappings from a tsconfig.json file can now be resolved.` line from the change-log?

asgerf replied:
I refactored the change note a bit, but for some reason the diff is really weird when looking at the commit alone. If you look at the change notes under "Files changed" you should be able to see this line occur a few lines earlier.

esbena commented on javascript/ql/src/semmle/javascript/dataflow/Configuration.qll:
```diff
@@ -633,7 +633,12 @@ private predicate exploratoryFlowStep(
   (cfg.isSource(nd) or nd.(AdditionalSource).isSourceFor(cfg)) and
-  lbl = FlowLabel::data()
+  (
+    if cfg instanceof TaintTracking::Configuration then
```
It is a shame that we have to introduce a dependency form `Configuration.qll` to `TaintTracking.qll` (I think this is the first, at least). We could avoid that by introducing `cfg.getDefaultSourceLabel()`, which in `TaintTracking::configuration` is implemented as `result = FlowLabel::taint()`. 
I will leave it to you to decide what to do here.

Ditto for `isSink` below.

asgerf replied:
That's very true, but on the other hand, there's also a downside to extending our public API with stuff that users don't actually need.

asgerf commented:
There was a test failure that revealed an interesting, but generally benign side effect of the change.

The prototype pollution utility query had a barrier guard for `x in obj` where `x` was sanitized in the _false_ case, while a standard sanitizer guard sanitizes `x` in the _true_ case. The latter one shouldn't matter because we don't use a taint-tracking configuration, but because its underlying instance (the `InExpr`) was included in `isBarrierGuard` the standard sanitizer was pulled in that way. Since the standard sanitizer guard only blocks the `taint` label now, it's still being pulled in, but has no effect as we're not using the `taint` label in that query.

The change thus fixed a spurious sanitizer which revealed that a test was passing for the wrong reason. I've updated the test accordingly.

asgerf commented:
This will need another evaluation. After rebasing some join orderings have changed and it's not performing as well anymore.

asgerf commented:
[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/js/property-barriers_1584665090114) still looks fine.

asgerf commented:
Thanks for the review @erik-krogh!

@esbena I believe I addressed your comment in in https://github.com/Semmle/ql/pull/2919/commits/b6ca4fbee33271f581ab5180e86749b3096ed885 but in any case, I've gone ahead and merged to unblock Erik's PR and since there are now a bunch of performance-related PRs floating around whose evaluations ought to be based on this one from now on.
